packages/gatsby/src/query/query-compiler.js

```diff
@@ -260,47 +213,110 @@ class Runner {
+        actions.queryExtractionGraphQLError({
           componentPath: filePath,
         })
-        return
+        continue
       }
-      let text
-      try {
-        text = filterContextForNode(printContext.getRoot(name), printContext)
-          .documents()
-          .map(GraphQLIRPrinter.print)
-          .join(`\n`)
-      } catch (error) {
-        const regex = /Unknown\sdocument\s`(.*)`/gm
-        const str = error.toString()
-        let m
-
-        let fragmentName
-        while ((m = regex.exec(str)) !== null) {
-          // This is necessary to avoid infinite loops with zero-width matches
-          if (m.index === regex.lastIndex) regex.lastIndex++
-
-          fragmentName = m[1]
-        }
-
-        const closestFragment = fragments
-          .map(f => {
-            return { fragment: f, score: levenshtein.get(fragmentName, f) }
-          })
-          .filter(f => f.score < 10)
-          .sort((a, b) => a.score > b.score)[0]?.fragment
 
-        addError({
-          id: `85908`,
-          filePath,
-          context: { fragmentName, closestFragment },
+      const usedFragments = new Set()
+      const stack = [operation]
+      let missingFragment = false
+
+      while (stack.length > 0) {
+        const def = stack.pop(operation)
+        visit(def, {
+          [Kind.FRAGMENT_SPREAD]: node => {
+            const name = node.name.value
+            if (usedFragmentsForFragment.has(name)) {
+              usedFragmentsForFragment
+                .get(name)
+                .forEach(derivedFragmentName => {
+                  usedFragments.add(derivedFragmentName)
+                })
+              usedFragments.add(name)
+            } else if (fragmentMap.has(name)) {
+              stack.push(fragmentMap.get(name).def)
+              usedFragments.add(name)
+            } else {
+              missingFragment = true
+              const closestFragment = fragmentNames
+                .map(f => {
+                  return { fragment: f, score: levenshtein.get(name, f) }
+                })
+                .filter(f => f.score < 10)
+                .sort((a, b) => a.score > b.score)[0]?.fragment
+
+              actions.queryExtractionGraphQLError({
+                componentPath: filePath,
+              })
+
+              addError({
+                id: `85908`,
+                filePath,
+                context: {
+                  fragmentName: name,
+                  closestFragment,
+                  codeFrame: codeFrameColumns(
+                    def.text,
+                    {
+                      start: getLocation({ body: def.text }, node.loc.start),
+                      end: getLocation({ body: def.text }, node.loc.end),
+                    },
+                    {
+                      linesAbove: 10,
+                      linesBelow: 10,
+                    }
+                  ),
+                },
+              })
+            }
+          },
```

Comment:
This is great! One issue though is that the `write` method becomes quite big and has multiple pieces of validations in it. Maybe we could extract some of them to separate methods to communicate the intent?

Reply:
Good idea.